ProjFS.Mac/PrjFSKext/FileSystemMountPoints.cpp

```diff
@@ -0,0 +1,131 @@
+#include "FileSystemMountPointsPrivate.hpp"
+#include "FileSystemMountPoints.hpp"
+#include "kernel-header-wrappers/mount.h"
+#include "Locks.hpp"
+#include "KextLog.hpp"
+#include <kern/assert.h>
+
+static Mutex s_mountPointsMutex;
+KEXT_STATIC MountPoint s_usedMountPoints[MaxUsedMountPoints];
+
+bool MountPoint_Init()
+{
+    s_mountPointsMutex = Mutex_Alloc();
+    if (!Mutex_IsValid(s_mountPointsMutex))
+    {
+        return false;
+    }
+    
+    return true;
+}
+
+void MountPoint_Cleanup()
+{
+    if (Mutex_IsValid(s_mountPointsMutex))
+    {
+        Mutex_FreeMemory(&s_mountPointsMutex);
+    }
+    
+    for (size_t i = 0; i < MaxUsedMountPoints; ++i)
+    {
+        // by the time this function is called, any provider user clients
+        // should have been terminated & deregistered
+        assert(s_usedMountPoints[i].mountPoint == nullptr);
+        assert(s_usedMountPoints[i].authCacheDisableCount == 0);
+    }
+}
+
+KEXT_STATIC ssize_t FindMountPoint(mount_t mountPoint)
+{
+    for (size_t i = 0; i < MaxUsedMountPoints; ++i)
+    {
+        if (s_usedMountPoints[i].mountPoint == mountPoint)
+        {
+            return i;
+        }
+    }
+    
+    return -1;
+}
+
+KEXT_STATIC_INLINE ssize_t FindEmptyMountPointSlot()
+{
+    return FindMountPoint(nullptr);
+}
+
+void MountPoint_DisableAuthCache(mount_t mountPoint)
+{
+    Mutex_Acquire(s_mountPointsMutex);
+    {
+        ssize_t mountPointIndex = FindMountPoint(mountPoint);
+        if (mountPointIndex < 0)
+        {
+            mountPointIndex = FindEmptyMountPointSlot();
+            if (mountPointIndex >= 0)
+            {
+                s_usedMountPoints[mountPointIndex].mountPoint = mountPoint;
+                assert(s_usedMountPoints[mountPointIndex].authCacheDisableCount == 0);
+            }
+        }
+        
+        if (mountPointIndex < 0)
```

Comment:
This would read easier for me if was added as an else for the above
`if (mountPointIndex < 0)`
followed by a return

Reply:
Do you mean:

        if (mountPointIndex < 0)
        {
            mountPointIndex = FindEmptyMountPointSlot();
            if (mountPointIndex >= 0)
            {
                s_usedMountPoints[mountPointIndex].mountPoint = mountPoint;
                assert(s_usedMountPoints[mountPointIndex].authCacheDisableCount == 0);
            }
            else
            {
               // No space in array, just disable auth cache and don't save previous state
               KextLog_Info("MountPoint_DisableAuthCache: No space to save auth cache state for mount point '%s', restoring will not be possible.", vfs_statfs(mountPoint)->f_mntonname);
               vfs_setauthcache_ttl(mountPoint, 0);
               Mutex_Release(s_mountPointsMutex); // <--- I don't really like this here
               return;
            }
        }

        MountPoint* usedMountPoint = &s_usedMountPoints[mountPointIndex];
        …

? I'm not a fan of breaking up the mutex release like that. Another option would be a `goto` (again, not a fan), or wrapping the whole function apart from the mutex lock/unlock in `do … while(false);` so that we're allowed to put a `break;` statement after the `vfs_setauthcache_ttl(mountPoint, 0);` call above.

Reply:
I've just pushed a new version of the code that moves this chunk of functionality into a new `TryFindOrInsertMountPoint_Locked()` helper function. This also allows us to easily unit test this code, so there are now a few tests too.